File: bench/core/bench-fanout.ts
// Steady-state per-request fan-out benchmark (TS side), mirroring the Rust
// `bench_fanout.rs`. Opens storage ONCE, registers K live subscribers, then
// times only the M `postGroupMessage` calls (1 insert + K fan-out pushes each)
// — the cost a real serving loop pays per post request. Subscribers drain
// concurrently (untimed) so the post loop measures insert+fan-out, not drain.
//
// Run:
//   NODE_PATH=<repo>/references/cordn/node_modules node bench/core/bench-fanout.ts
// Env: CORDN_BENCH_FANOUT (8), CORDN_BENCH_MESSAGES (1000),
//      CORDN_BENCH_BACKEND (both | sqlite | memory)

import { Coordinator } from "../../references/cordn/src/coordinator/coordinator.ts";
import { InMemoryCoordinatorStorage } from "../../references/cordn/src/coordinator/storage/inMemoryStorage.ts";
import { SqliteCoordinatorStorage } from "../../references/cordn/src/coordinator/storage/sqliteStorage.ts";
import { createPrivateMessage } from "../../references/cordn/src/coordinator/testUtils.ts";

function readPositiveIntEnv(name, fallback) {
  const raw = process.env[name];
  if (!raw) {
    return fallback;
  }
  const parsed = Number.parseInt(raw, 10);
  if (!Number.isFinite(parsed) || parsed <= 0) {
    throw new Error(`Environment variable ${name} must be a positive integer`);
  }
  return parsed;
}

function readBackends() {
  switch (process.env.CORDN_BENCH_BACKEND) {
    case "sqlite":
      return ["sqlite"];
    case "memory":
      return ["memory"];
    default:
      return ["sqlite", "memory"];
  }
}

async function consumeRecords(messages, expectedCount) {
  const iterator = messages[Symbol.asyncIterator]();
  try {
    for (let index = 0; index < expectedCount; index += 1) {
      const result = await iterator.next();
      if (result.done) {
        throw new Error(`ended early after ${index}, expected ${expectedCount}`);
      }
    }
  } finally {
    await iterator.return?.();
  }
}

async function runBackend(backend, fanout, messages) {
  const storage =
    backend === "memory"
      ? new InMemoryCoordinatorStorage()
      : new SqliteCoordinatorStorage({ path: ":memory:" });
  const coordinator = new Coordinator({ storage });

  const subscriptions = Array.from({ length: fanout }, () =>
    coordinator.subscribeGroupMessages({ groupId: "g" }),
  );

  // K concurrent drainers; suspended at their first `await .next()` so they
  // don't run during the synchronous post loop below.
  const drainers = subscriptions.map((subscription) =>
    consumeRecords(subscription.messages, messages),
  );

  const start = process.hrtime.bigint();
  for (let index = 0; index < messages; index += 1) {
    coordinator.postGroupMessage({
      groupId: "g",
      opaqueMessage: createPrivateMessage({
        groupId: "g",
        epoch: 1n,
        contentType: 1,
        bytes: [index % 251],
      }),
    });
  }
  const totalNs = Number(process.hrtime.bigint() - start);

  // Untimed drain + correctness (each subscriber received all `messages`).
  await Promise.all(drainers);
  storage.close();

  const totalMs = totalNs / 1_000_000;
  const secs = totalNs / 1_000_000_000;
  return {
    backend,
    totalMs,
    avgUs: (totalMs * 1000) / messages,
    msgsPerSec: messages / secs,
    deliveriesPerSec: (messages * fanout) / secs,
  };
}

async function main() {
  const fanout = readPositiveIntEnv("CORDN_BENCH_FANOUT", 8);
  const messages = readPositiveIntEnv("CORDN_BENCH_MESSAGES", 1000);

  console.log("TypeScript cordn — steady-state fan-out benchmark");
  console.log(`fanout=${fanout} messages=${messages}`);
  console.log("");

  for (const backend of readBackends()) {
    const r = await runBackend(backend, fanout, messages);
    console.log(`── ${backend} ──────────────────────────────`);
    console.log(
      `  total=${r.totalMs.toFixed(2)}ms  avg=${r.avgUs.toFixed(
        3,
      )}µs/post  msgs/sec=${r.msgsPerSec.toFixed(0)}  deliveries/sec=${r.deliveriesPerSec.toFixed(
        0,
      )}`,
    );
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: crates/cordn-core/examples/bench_fanout.rs
//! Steady-state per-request fan-out benchmark. Unlike `bench_subscriptions`,
//! the storage is opened ONCE and only the per-message server work is measured:
//! each `post_group_message` is 1 storage insert + a fan-out push to K live
//! subscribers — the cost a real serving loop pays per post request (not a
//! bulk-seed stress). K subscriber tasks drain concurrently (untimed) so the
//! post loop measures insert+fan-out, not drain.
//!
//! Run: cargo run --release -p cordn-core --example bench_fanout
//! Env: CORDN_BENCH_FANOUT (8)   — live subscribers per group
//!      CORDN_BENCH_MESSAGES (1000) — messages posted in the measured window
//!      CORDN_BENCH_BACKEND (both | sqlite | memory)

use std::env;
use std::sync::Arc;
use std::time::Instant;

use cordn_core::{
    Coordinator, CoordinatorOptions, CoordinatorStorage, InMemoryCoordinatorStorage,
    PostGroupMessageInput, SqliteCoordinatorStorage,
};

#[derive(Clone, Copy, PartialEq, Eq)]
enum Backend {
    Sqlite,
    Memory,
}

impl Backend {
    fn label(self) -> &'static str {
        match self {
            Backend::Sqlite => "SQLite",
            Backend::Memory => "In-memory",
        }
    }

    fn build(self) -> Arc<dyn CoordinatorStorage> {
        match self {
            Backend::Sqlite => Arc::new(
                SqliteCoordinatorStorage::open_in_memory()
                    .expect("open in-memory sqlite for bench"),
            ),
            Backend::Memory => Arc::new(InMemoryCoordinatorStorage::new()),
        }
    }
}

fn read_backends() -> Vec<Backend> {
    match env::var("CORDN_BENCH_BACKEND").ok().as_deref() {
        Some("sqlite") => vec![Backend::Sqlite],
        Some("memory") => vec![Backend::Memory],
        _ => vec![Backend::Sqlite, Backend::Memory],
    }
}

fn read_pos_env(name: &str, fallback: usize) -> usize {
    match env::var(name) {
        Ok(raw) => raw
            .trim()
            .parse::<usize>()
            .ok()
            .filter(|v| *v > 0)
            .unwrap_or_else(|| panic!("{name} must be a positive integer, got {raw:?}")),
        Err(_) => fallback,
    }
}

// ponytail: fixed-size synthetic payload; revisit if results look size-sensitive.
fn opaque_message(index: usize) -> Vec<u8> {
    let size = std::env::var("CORDN_BENCH_MSG_BYTES")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(32)
        .max(1);
    let mut bytes = vec![0u8; size];
    let last = bytes.len() - 1;
    bytes[last] = (index % 251) as u8;
    bytes
}

struct FanoutResult {
    backend: Backend,
    total_ms: f64,
    avg_us: f64,
    msgs_per_sec: f64,
    deliveries_per_sec: f64,
}

/// Measure steady-state post+fan-out for one backend. Opens storage once,
/// registers K live-tail subscribers, then times only the M post calls.
async fn run_backend(backend: Backend, fanout: usize, messages: usize) -> FanoutResult {
    let coord = Coordinator::new(CoordinatorOptions {
        storage: Some(backend.build()),
        cleanup_interval_ms: Some(0),
        ..CoordinatorOptions::default()
    });

    let subs = (0..fanout)
        .map(|_| coord.subscribe_group_messages("g"))
        .collect::<Vec<_>>();

    // K concurrent drainers; on the current_thread runtime they only progress
    // once we await (after the timed post loop), so the post loop measures
    // insert + fan-out pushes, exactly like the TS side's synchronous post loop.
    let handles = subs
        .into_iter()
        .map(|mut sub| {
            tokio::spawn(async move {
                for _ in 0..messages {
                    sub.recv().await.expect("drain message");
                }
            })
        })
        .collect::<Vec<_>>();

    let start = Instant::now();
    for i in 0..messages {
        coord
            .post_group_message(PostGroupMessageInput {
                group_id: "g".into(),
                opaque_message: opaque_message(i),
            })
            .expect("post");
    }
    let elapsed = start.elapsed();

    // Untimed drain + correctness (each drainer received all `messages`).
    for handle in handles {
        handle.await.expect("drainer task");
    }
    coord.close();

    let secs = elapsed.as_secs_f64();
    FanoutResult {
        backend,
        total_ms: secs * 1000.0,
        avg_us: secs * 1_000_000.0 / messages as f64,
        msgs_per_sec: messages as f64 / secs,
        deliveries_per_sec: (messages * fanout) as f64 / secs,
    }
}

// current_thread mirrors Node's single-threaded event loop.
#[tokio::main(flavor = "current_thread")]
async fn main() {
    let fanout = read_pos_env("CORDN_BENCH_FANOUT", 8);
    let messages = read_pos_env("CORDN_BENCH_MESSAGES", 1000);

    println!("Rust cordn-core — steady-state fan-out benchmark");
    println!("fanout={fanout} messages={messages}");
    println!();

    for backend in read_backends() {
        let r = run_backend(backend, fanout, messages).await;
        println!("── {} ──────────────────────────────", r.backend.label());
        println!(
            "  total={:.2}ms  avg={:.3}µs/post  msgs/sec={:.0}  deliveries/sec={:.0}",
            r.total_ms, r.avg_us, r.msgs_per_sec, r.deliveries_per_sec,
        );
    }
}
